src/peft/tuners/vera/bnb.py

```diff
@@ -0,0 +1,408 @@
+                )
+                vera_data = self.get_delta_weight(active_adapter)
+
+                weight = self.get_base_layer().weight
+                kwargs = weight.__dict__
+                w_data = bnb.functional.dequantize_4bit(weight.data, weight.quant_state) - vera_data
+
+                self.get_base_layer().weight = bnb.nn.Params4bit(w_data.to("cpu"), requires_grad=False, **kwargs).to(
+                    weight.device
+                )
+
+        def get_delta_weight(self, adapter) -> torch.Tensor:
+            vera_A = self.vera_A[adapter]
+            vera_B = self.vera_B[adapter]
+
+            device = vera_B.device
+            dtype = vera_B.dtype
+
+            cast_to_fp32 = device.type == "cpu" and (dtype == torch.float16 or dtype == torch.bfloat16)
+
+            lambda_d = self.vera_lambda_d[adapter]
+            lambda_b = self.vera_lambda_b[adapter]
+
+            if cast_to_fp32:
+                vera_A = vera_A.float()
+                vera_B = vera_B.float()
+                lambda_d = lambda_d.float()
+                lambda_b = lambda_b.float()
+
+            sliced_A = vera_A[:, : self.in_features]
+            sliced_B = vera_B[: self.out_features, :]
+            lambda_b = lambda_b.unsqueeze(-1)
+            lambda_d = lambda_d.unsqueeze(-1)
+
+            output_tensor = transpose((lambda_b * sliced_B) @ (lambda_d * sliced_A), self.fan_in_fan_out)
+
+            if cast_to_fp32:
+                output_tensor = output_tensor.to(dtype=dtype)
+
+            return output_tensor
+
+        def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+            if self.disable_adapters:
+                if self.merged:
+                    self.unmerge()
+                result = self.base_layer(x, *args, **kwargs)
+            elif self.merged:
+                result = self.base_layer(x, *args, **kwargs)
+            else:
+                result = self.base_layer(x, *args, **kwargs)
+                result = result.clone()
+                for active_adapter in self.active_adapters:
+                    if active_adapter not in self.vera_lambda_d.keys():
+                        continue
+
+                    lambda_d = self.vera_lambda_d[active_adapter]
+                    lambda_b = self.vera_lambda_b[active_adapter]
+
+                    vera_A = self.vera_A[active_adapter]
+                    vera_B = self.vera_B[active_adapter]
+
+                    dropout = self.vera_dropout[active_adapter]
+
+                    requires_conversion = not torch.is_autocast_enabled()
+                    if requires_conversion:
+                        expected_dtype = result.dtype
+                        compute_dtype = lambda_d.dtype
+                        if x.dtype != compute_dtype:
+                            x = x.to(compute_dtype)
+
+                    sliced_A = vera_A.repeat(1, self.in_features)
```

Comment:
The way that the original layer's shape is determined no longer works when bnb layers are used. Specifically, these lines:

https://github.com/huggingface/peft/blob/f4cf170a9c51d822f950cde0a0e1c87dc013403a/src/peft/tuners/vera/model.py#L122-L125

Could you try replacing those lines with these instead:

```python
            if isinstance(module, nn.Linear):
                module_shape = module.in_features, module.out_features
            elif isinstance(module, Conv1D):
                module_shape = module.weight.ds_shape if hasattr(module.weight, "ds_shape") else module.weight.shape
```

This should allow you to use the normal `forward` code for 4 bit, since the `repeat` call, as you mentioned, is incorrect.